File: go/epod.go
package zymeup

import (
	"encoding/json"
	"fmt"
	"net/url"
)

type EpodClient struct {
	client *Client
}

func NewEpodClient(c *Client) *EpodClient {
	return &EpodClient{client: c}
}

type EpodListItem struct {
	ID          string `json:"id"`
	TrackingNo  string `json:"tracking_no"`
	Status      string `json:"status"`
	RecipientName string `json:"recipient_name,omitempty"`
	CreatedAt   string `json:"created_at"`
}

type EpodListResponse struct {
	Data       []EpodListItem `json:"data"`
	Total      int            `json:"total"`
	Page       int            `json:"page"`
	PageSize   int            `json:"page_size"`
}

type EpodDetail struct {
	ID             string `json:"id"`
	TrackingNo     string `json:"tracking_no"`
	Status         string `json:"status"`
	RecipientName  string `json:"recipient_name,omitempty"`
	SignURL        string `json:"sign_url,omitempty"`
	EvidenceHash   string `json:"evidence_hash,omitempty"`
	SignatureData  string `json:"signature_data,omitempty"`
	PhotoURL       string `json:"photo_url,omitempty"`
	ProofType      string `json:"proof_type,omitempty"`
	CreatedAt      string `json:"created_at"`
	UpdatedAt      string `json:"updated_at"`
}

type SignURLResponse struct {
	SignURL string `json:"sign_url"`
}

func (e *EpodClient) List(page, pageSize int, status string) (*EpodListResponse, error) {
	params := url.Values{}
	if page > 0 {
		params.Set("page", fmt.Sprintf("%d", page))
	}
	if pageSize > 0 {
		params.Set("page_size", fmt.Sprintf("%d", pageSize))
	}
	if status != "" {
		params.Set("status", status)
	}
	path := "/api/v1/shipment/epod/list?" + params.Encode()
	resp, err := e.client.doRequest("GET", path, nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int             `json:"code"`
		Data EpodListResponse `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result.Data, nil
}

func (e *EpodClient) Get(id string) (*EpodDetail, error) {
	resp, err := e.client.doRequest("GET", "/api/v1/shipment/epod/"+id, nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int        `json:"code"`
		Data EpodDetail `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result.Data, nil
}

func (e *EpodClient) GenerateSignURL(id string) (*SignURLResponse, error) {
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/"+id+"/sign", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int             `json:"code"`
		Data SignURLResponse `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result.Data, nil
}

func (e *EpodClient) GeneratePdf(id string) (map[string]interface{}, error) {
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/"+id+"/pdf", nil)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (e *EpodClient) Verify(id string) (bool, error) {
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/"+id+"/verify", nil)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int  `json:"code"`
		Data struct {
			Verified bool `json:"verified"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return false, err
	}
	return result.Data.Verified, nil
}

func (e *EpodClient) CaptureProof(id string, data map[string]interface{}) (map[string]interface{}, error) {
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/"+id+"/capture-proof", data)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (e *EpodClient) Create(data map[string]interface{}) (map[string]interface{}, error) {
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/create", data)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (e *EpodClient) GenerateFromOrder(orderID string, options map[string]interface{}) (map[string]interface{}, error) {
	body := map[string]interface{}{"order_id": orderID}
	for k, v := range options {
		body[k] = v
	}
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/generate-from-order", body)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (e *EpodClient) Update(id string, data map[string]interface{}) (map[string]interface{}, error) {
	resp, err := e.client.doRequest("PUT", "/api/v1/shipment/epod/"+id+"/update", data)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (e *EpodClient) Deliver(id string, data map[string]interface{}) (map[string]interface{}, error) {
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/"+id+"/delivery", data)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

func (e *EpodClient) Fail(id string, data map[string]interface{}) (map[string]interface{}, error) {
	resp, err := e.client.doRequest("POST", "/api/v1/shipment/epod/"+id+"/fail", data)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var result struct {
		Code int                    `json:"code"`
		Data map[string]interface{} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Data, nil
}

File: node/src/epod.ts
import { HttpClient } from './http-client';
import type { ShipzyConfig } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export interface EpodListItem {
    id: string;
    tracking_no: string;
    status: string;
    recipient_name?: string;
    created_at: string;
}

export interface EpodListResponse {
    data: EpodListItem[];
    total: number;
    page: number;
    page_size: number;
}

export interface EpodDetail {
    id: string;
    tracking_no: string;
    status: string;
    recipient_name?: string;
    recipient_phone?: string;
    delivery_address?: Record<string, any>;
    sender_address?: Record<string, any>;
    proof_type?: string;
    created_at: string;
    updated_at: string;
    sign_url?: string;
    evidence_hash?: string;
    document_hash?: string;
    signature_data?: string;
    photo_url?: string;
}

export interface SignUrlResponse {
    sign_url: string;
}

// ============ Client ============

export class EpodClient extends HttpClient {
    async list(params: { page?: number; pageSize?: number; status?: string; trackingNo?: string } = {}): Promise<ApiResult<EpodListResponse>> {
        const q = this.buildQuery({ page: params.page, page_size: params.pageSize, status: params.status, tracking_no: params.trackingNo });
        return this.request(`/api/v1/shipment/epod/list${q}`);
    }

    async get(id: string): Promise<ApiResult<EpodDetail>> {
        return this.request(`/api/v1/shipment/epod/${id}`);
    }

    async create(data: Record<string, unknown>): Promise<ApiResult<EpodDetail>> {
        return this.request('/api/v1/shipment/epod/create', 'POST', data);
    }

    async generateFromOrder(orderId: string, options: Record<string, unknown> = {}): Promise<ApiResult<EpodDetail>> {
        return this.request('/api/v1/shipment/epod/generate-from-order', 'POST', { order_id: orderId, ...options });
    }

    async update(id: string, data: Record<string, unknown>): Promise<ApiResult<EpodDetail>> {
        return this.request(`/api/v1/shipment/epod/${id}/update`, 'PUT', data);
    }

    async deliver(id: string, data: Record<string, unknown> = {}): Promise<ApiResult<SignUrlResponse & { sign_token_expires_at?: string }>> {
        return this.request(`/api/v1/shipment/epod/${id}/delivery`, 'POST', data);
    }

    async fail(id: string, data: { remark: string }): Promise<ApiResult<EpodDetail>> {
        return this.request(`/api/v1/shipment/epod/${id}/fail`, 'POST', data);
    }

    async captureProof(id: string, data: Record<string, unknown>): Promise<ApiResult<EpodDetail>> {
        return this.request(`/api/v1/shipment/epod/${id}/capture-proof`, 'POST', data);
    }

    async verify(id: string): Promise<ApiResult<{ verified: boolean; error?: string }>> {
        return this.request(`/api/v1/shipment/epod/${id}/verify`, 'POST', {});
    }

    async generateSignUrl(id: string): Promise<ApiResult<SignUrlResponse>> {
        return this.request(`/api/v1/shipment/epod/${id}/sign`, 'POST', {});
    }

    async generatePdf(id: string): Promise<ApiResult<{ pdf_url?: string }>> {
        return this.request(`/api/v1/shipment/epod/${id}/pdf`, 'POST', {});
    }

    async uploadPhoto(id: string, file: File): Promise<ApiResult<{ photo_url: string }>> {
        const formData = new FormData();
        formData.append('file', file);
        return this.request(`/api/v1/shipment/epod/${id}/upload-photo`, 'POST', formData, true) as Promise<ApiResult<{ photo_url: string }>>;
    }
}
